core/napi.rs

```diff
@@ -0,0 +1,444 @@
+pub const napi_biguint64_array: napi_typedarray_type = 10;
+
+pub struct napi_type_tag {
+  pub lower: u64,
+  pub upper: u64,
+}
+
+pub type napi_callback =
+  unsafe extern "C" fn(env: napi_env, info: napi_callback_info) -> napi_value;
+
+pub type napi_finalize = unsafe extern "C" fn(
+  env: napi_env,
+  data: *mut c_void,
+  finalize_hint: *mut c_void,
+);
+
+pub type napi_async_execute_callback =
+  unsafe extern "C" fn(env: napi_env, data: *mut c_void);
+
+pub type napi_async_complete_callback =
+  unsafe extern "C" fn(env: napi_env, status: napi_status, data: *mut c_void);
+
+pub type napi_threadsafe_function_call_js = unsafe extern "C" fn(
+  env: napi_env,
+  js_callback: napi_value,
+  context: *mut c_void,
+  data: *mut c_void,
+);
+
+pub type napi_async_cleanup_hook =
+  unsafe extern "C" fn(env: napi_env, data: *mut c_void);
+
+pub type napi_property_attributes = i32;
+
+pub const napi_default: napi_property_attributes = 0;
+pub const napi_writable: napi_property_attributes = 1 << 0;
+pub const napi_enumerable: napi_property_attributes = 1 << 1;
+pub const napi_configurable: napi_property_attributes = 1 << 2;
+pub const napi_static: napi_property_attributes = 1 << 10;
+pub const napi_default_method: napi_property_attributes =
+  napi_writable | napi_configurable;
+pub const napi_default_jsproperty: napi_property_attributes =
+  napi_enumerable | napi_configurable | napi_writable;
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug)]
+pub struct napi_property_descriptor {
+  pub utf8name: *const c_char,
+  pub name: napi_value,
+  pub method: napi_callback,
+  pub getter: napi_callback,
+  pub setter: napi_callback,
+  pub value: napi_value,
+  pub attributes: napi_property_attributes,
+  pub data: *mut c_void,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct napi_extended_error_info {
+  pub error_message: *const c_char,
+  pub engine_reserved: *mut c_void,
+  pub engine_error_code: i32,
+  pub status_code: napi_status,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct napi_node_version {
+  pub major: u32,
+  pub minor: u32,
+  pub patch: u32,
+  pub release: *const c_char,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+/// Env that is shared between all contexts in same native module.
+pub struct EnvShared {
+  pub instance_data: *mut c_void,
+  pub data_finalize: Option<napi_finalize>,
+  pub data_finalize_hint: *mut c_void,
+  pub napi_wrap: v8::Global<v8::Private>,
+  pub finalize: Option<napi_finalize>,
+  pub finalize_hint: *mut c_void,
+  pub filename: *const c_char,
+}
+
+impl EnvShared {
+  pub fn new(napi_wrap: v8::Global<v8::Private>) -> Self {
+    Self {
+      instance_data: std::ptr::null_mut(),
+      data_finalize: None,
+      data_finalize_hint: std::ptr::null_mut(),
+      napi_wrap,
+      finalize: None,
+      finalize_hint: std::ptr::null_mut(),
+      filename: std::ptr::null(),
+    }
+  }
+}
+
+#[repr(C)]
+// #[derive(Debug)]
+pub struct Env<'a, 'b, 'c> {
+  pub scope: &'a mut v8::ContextScope<'b, v8::HandleScope<'c>>,
```

Comment:
This should store the isolate pointer too. We need it for `napi_open_handle_scope` when the scope is NULL (in completion callbacks for thread-safe functions)

Reply:
https://github.com/denoland/rusty_v8/pull/901